## ----------------------------------------------------Section - 3.2-----------------------------------------------------------
###                              3.2) Coventional Machine Learning Model - Support Vector Machines

In [5]:
import os
#import cv2
import pandas as pd
import numpy as np
#import tensorflow as tf
import matplotlib.pyplot as plt
import matplotlib as mpl

from os.path import join
#from tensorflow.keras import layers, models

In [ ]:
Step-2: Loading the Training, Test DataSet and exploring the data

In [7]:
# Training data reading
hair_colours = ['black', 'blond', 'brown', 'gray']
dataset_directory = 'C:\\Users\\suhas\\Documents\\study\\ML Practice\\image_dataset'  ## CHANGE TO YOUR OWN DIRECTORY

images_train = np.load(join(dataset_directory, 'train_images.npy'))
labels_train = np.load(join(dataset_directory, 'train_labels.npy'))

image_size = 48
images_train_rs = images_train.reshape([-1, image_size, image_size, 3]) / 255.0

In [3]:
Features = images_train_rs.shape[1]*images_train_rs.shape[2]*images_train_rs.shape[3]
Features

6912

In [4]:
# reshaping training data
lst1 = [None] * images_train_rs.shape[0]
for i in range(images_train_rs.shape[0]):
    lst1[i] = images_train_rs[i].reshape([-1])
#print(lst)
x = np.array(lst1).reshape(images_train_rs.shape[0],Features)
print(x.shape)

(98479, 6912)


##### Observations From preliminary data Exploration:
so we can see that images_train contains a list of size 98479 with 48 features as height for each picture, 48 features as width for each picture and a 3 channel (RGB) colour sclae for each picture (with values raning from 1 to 255). This results in 48*48*3 = 6912 features to work with.

In [5]:
# reading the test data
images_test_public = np.load(join(dataset_directory, 'test_public_images.npy'))
files_test_public = np.load(join(dataset_directory, 'test_public_files.npy'))
images_test_public_rs = images_test_public.reshape([-1, image_size, image_size, 3]) / 255.0

In [6]:
Features3 = images_test_public_rs.shape[1]*images_test_public_rs.shape[2]*images_test_public_rs.shape[3]
Features3

6912

In [7]:
lst4 = [None] * images_test_public_rs.shape[0]
for i in range(images_test_public_rs.shape[0]):
    lst4[i] = images_test_public_rs[i].reshape([-1])
#print(lst)
z = np.array(lst4).reshape(images_test_public_rs.shape[0],Features3)
print(z.shape)

(11921, 6912)


In [8]:
# reading the validation data set
images_val = np.load(join(dataset_directory, 'val_images.npy'))
labels_val = np.load(join(dataset_directory, 'val_labels.npy'))
files_val = np.load(join(dataset_directory, 'val_files.npy'))

print("Before reshaping, image val shape is:\n",images_val.shape)
print("Data inside images val:\n", images_val)
print("Data inside labels val:\n", labels_val)
print("Data inside files val:\n", files_val)
print("\n")

print("list features for a row, for example first picture features are:", images_val[0]) 
print("no of features for first picture are:", len(images_val[0]))
print("\n")

print("list of features inside the first feature for picture 1:\n", images_val[0][0])
print("no of features inside the first feature for picture 1:\n", len(images_val[0][0]))

images_val_rs = images_val.reshape([-1, image_size, image_size, 3]) / 255.0

Before reshaping, image val shape is:
 (12270, 48, 48, 3)
Data inside images val:
 [[[[236 240 243]
   [240 244 247]
   [249 252 255]
   ...
   [192 184 182]
   [183 182 178]
   [182 180 179]]

  [[236 240 243]
   [240 244 247]
   [249 252 255]
   ...
   [192 184 182]
   [183 182 178]
   [182 180 179]]

  [[236 240 243]
   [240 244 247]
   [249 252 255]
   ...
   [193 184 182]
   [183 182 178]
   [182 180 179]]

  ...

  [[237 239 240]
   [236 238 240]
   [235 237 239]
   ...
   [ 14  15  14]
   [ 18  18  17]
   [ 33  33  32]]

  [[235 240 241]
   [238 243 244]
   [246 251 252]
   ...
   [ 30  30  30]
   [ 28  28  28]
   [ 23  23  23]]

  [[239 241 243]
   [217 219 221]
   [141 143 144]
   ...
   [ 27  27  27]
   [ 27  27  27]
   [ 25  25  25]]]


 [[[ 98  88  81]
   [ 59  41  34]
   [ 69  44  39]
   ...
   [ 53  33  24]
   [ 54  34  25]
   [ 54  34  25]]

  [[ 98  88  81]
   [ 59  41  34]
   [ 69  44  39]
   ...
   [ 53  33  24]
   [ 54  34  25]
   [ 54  34  25]]

  [[ 99  89  82]
   

In [9]:
Features1 = images_val_rs.shape[1]*images_val_rs.shape[2]*images_val_rs.shape[3]
Features1

6912

In [10]:
lst3 = [None] * images_val_rs.shape[0]
for i in range(images_val_rs.shape[0]):
    lst3[i] = images_val_rs[i].reshape([-1])
#print(lst)
y = np.array(lst3).reshape(images_val_rs.shape[0],Features1)
print(y.shape)

(12270, 6912)


In [11]:
from sklearn.svm import SVC
model = SVC(random_state = 26)
model = model.fit(x, labels_train)
Val_pred = model.predict(y)
Test_pred = model.predict(z)

In [12]:
df1 = pd.DataFrame(Test_pred, columns=['Prediction'])
df1.tail() # creating a data frame for predicted values

Prediction
11916           1
11917           1
11918           0
11919           0
11920           2

In [13]:
df2 = pd.DataFrame(files_test_public, columns=['ID'])
df2.tail() # creating a data frame for files_val which contains image names

ID
11916  202595.jpg
11917  202596.jpg
11918  202597.jpg
11919  202598.jpg
11920  202599.jpg

In [14]:
# Joining the two data frames through the index
Test_Public_pred_Acc_testpurpose_45693242 = df2.merge(df1, left_index=True, right_index=True)

In [15]:
Test_Public_pred_Acc_testpurpose_45693242.head()

ID  Prediction
0  182639.jpg           0
1  182640.jpg           2
2  182642.jpg           0
3  182643.jpg           0
4  182644.jpg           0

In [16]:
Test_Public_pred_Acc_testpurpose_45693242.to_csv (r'C:\\Users\\suhas\\Downloads\\ML Practice\\SVM_default_fulldata_45693242.csv', index = False, header=True)

In [11]:
from pickle import dump

In [13]:
import pickle

In [18]:
# saving the model for reusing purpose
dump(model, open('model1.pkl', 'wb'))

In [16]:
# using the saved model to predict on some 5 values of validation data set
load_model = pickle.load(open("model1.pkl", 'rb'))
Val_pred = load_model.predict(y[:5])

In [17]:
Val_pred

array([0, 2, 2, 0, 0], dtype=int64)